rfcs/2021-12-01-9460-health-endpoint-improvements.md

```diff
@@ -0,0 +1,117 @@
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+The API will not start until the topology is build and validated as functional,
+this isn't precisely when Vector's configured sources _can_ actually process
+events but it's a reasonable first step to ensure graceful startups.
+
+### Implementation
+
+We have existing logic in place for other components to handle Vector's shutdown.
+The [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L7)
+can be updated to respond 503 if shutdown has started and 200 otherwise. This
+should not negatively impact existing deployments.
+
+## Rationale
+
+This improvement is low effort but greatly improves the experience of running
+Vector behind a load balancer or in a service discovery system. As we recommend
+both of those options for production deployments there's very little reason not
+to implement this.
+
+## Drawbacks
+
+N/A
+
+## Prior Art
+
+- [Beats](https://www.elastic.co/guide/en/beats/filebeat/7.15/http-endpoint.html)
+  - [Liveness and Readiness Probes](https://github.com/elastic/helm-charts/blob/715eeda8a45b8c3d8542921f5485aa502c238d93/filebeat/values.yaml#L174-L198)
+- [FluentBit](https://docs.fluentbit.io/manual/administration/monitoring#rest-api-interface)
+  - [Liveness and Readiness Probes](https://github.com/fluent/helm-charts/blob/355575c5b2a5bd858bcadeaa9d8d5d7f15a7816d/charts/fluent-bit/values.yaml#L132-L140)
+- [Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/cluster-health.html)
+  - [Readiness Probe](https://github.com/elastic/helm-charts/blob/715eeda8a45b8c3d8542921f5485aa502c238d93/elasticsearch/templates/statefulset.yaml#L227-L291)
+- [Datadog Agent]
+  - [Liveness and Readiness Probes](https://github.com/DataDog/helm-charts/blob/d5e1f4370442bdc5e457468ac7ff0ff943f528d5/charts/datadog/templates/_container-agent.yaml#L193-L199)
+  - [agent health](https://docs.datadoghq.com/agent/guide/agent-commands/?tab=agentv6v7#other-commands)
+
+## Alternatives
+
+### Do nothing
+
+Not adding additional responses to the existing health endpoint seems strictly
+worse, as we don't have any ways to stop routing traffic to instances that are
+shutting down. Those requests will already be rejected by sources and this
+behaviour reflects poorly on zero downtime deployments, even though events
+shouldn't be dropped.
+
+### `vector health` subcommand
+
+While not suitable for general load balancer usage, it's very easy to `exec`
+commands in Kubernetes to determine health. For prior art we don't have to look
+further than the Datadog Agent which has a subcommand that outputs current
+health information that can be used by operators/systems.
+
+## Outstanding Questions
+
+- Do we need to update the `{ health }` GraphQL query at the same time?
```

Comment:
Yes, shouldn't both endpoints use the same code path?

Reply:
I do not believe they currently are, and regardless I don't think they need to serve the same purpose. If they do serve the same purpose I'd argue that one of them should be removed entirely. cc @leebenson

Reply:
I'm okay with leaving `{ health }` in the API as-is, and tackling later if needed.

Reply:
I'd prefer to remove it if they are going to deviate in behavior. I can't think of a use case for that GraphQL query anyway. Typically this is used for load balancing purposes.

Reply:
@binarylogic referencing @leebenson's thoughts on this previously: https://github.com/vectordotdev/vector/pull/10297#discussion_r769672177

Reply:
Do we actually make use of the health query in the `top` / `tap` clients:

https://github.com/vectordotdev/vector/blob/7f8e41c026a4b24b81d21984766ffd423b1b38a8/lib/vector-api-client/src/client.rs#L30-L49

Reply:
(That last comment should have read "We do", not "Do we". A statement, not a question 😄 )